enhancements/machine-config/on-cluster-builds.md

```diff
@@ -0,0 +1,1009 @@
+  IsSuccess --> | Yes | DeleteConfigMaps("Delete ConfigMaps")
+  IsSuccess --> | No | LeaveConfigMaps("Leave ConfigMaps \n for inspection")
+  DeleteConfigMaps --> Stop
+  LeaveConfigMaps --> Stop
+```
+
+Once the bootc ConfigMap annotation standards are better known, we could
+potentially skip the gzip / base64-encoding step by converting MachineConfigs
+into multiple ConfigMaps, temporarily storing them in the API server, injecting
+them into the build context, then deleting them once the build is complete.
+
+Within the build context itself, we'd need a different mechanism than Ignition
+live-apply to consume those ConfigMaps and inject their contents into the
+appropriate places within the image being built. Additionally, one of the big
+ideas around the bootc ConfigMap spec is using annotations to indicate what
+directory to store the files in, what file modes they should have, etc. This
+data would be unavailable via a volume mount. So it may be necessary to create a
+map of ConfigMap name -> bootc annotations and store this in an additional
+ConfigMap. This would also necessitate the need for a binary such as Ignition
+live-apply to consume this information from within the build context to write
+those files into the appropriate places.
+
+This could look like this instead:
+
+```mermaid
+flowchart TB
+  Start("Start")
+  subgraph Prepare [Prepare For Build]
+    direction TB
+    RenderDockerfile("Render Dockerfile")
+    PrepMachineConfig("Convert into bootc ConfigMaps")
+    PrepIndexConfigMap("Prepare bootc index ConfigMap")
+    ConfigMaps("Create ephemeral ConfigMaps")
+  
+    RenderDockerfile --> ConfigMaps
+    PrepMachineConfig --> PrepIndexConfigMap --> ConfigMaps
+  end
+  
+  subgraph ImageBuilder [Image Builder Context]
+    direction TB
+    ParseIndexConfigMap("Parse bootc index ConfigMap")
+    ApplyConfigMapsToImage("Apply bootc ConfigMaps \n to image build")
+    RunAdminProvidedContent("Run admin-provided \n Dockerfile content")
+    ParseIndexConfigMap-->ApplyConfigMapsToImage-->RunAdminProvidedContent
+  end
+
+  Stop("Stop")
+
+  Start --> Prepare
+  ConfigMaps --> |Mounted as volumes into image build context| ImageBuilder
+  IsSuccess{"Is Build Successful?"}
+
+  ImageBuilder --> IsSuccess
+  IsSuccess --> | Yes | DeleteConfigMaps("Delete ConfigMaps")
+  IsSuccess --> | No | LeaveConfigMaps("Leave ConfigMaps \n for inspection")
+  DeleteConfigMaps --> Stop
+  LeaveConfigMaps --> Stop
+```
+
+In terms of the ConfigMaps, they would look something like this:
+```yaml
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: etc-hello-world
+  annotations:
+    bootc.coreos.io/baseDir: /etc
+data:
+  hello-world: "hello world"
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: etc-ohai
+  annotations:
+    bootc.coreos.io/baseDir: /etc
+data:
+  ohai: "O Hai!"
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: authorized-ssh-keys
+  annotations:
+    bootc.coreos.io/baseDir: /home/core/.ssh
```

Comment:
Given all the annotation that we have to add per field, a custom CRD approach may be cleaner here.

Reply:
I agree. But I think this may have a hard requirement on this data being represented as ConfigMaps. Here's why:
- To my understanding, bootc will only understand ConfigMaps (at least right now).
- Pods and OpenShift Image Builds can accept ConfigMaps to use as a volume source.

It may be OK in this particular context, these would be created by the BuildController, mounted into the build context, their contents consumed by a bootc-like process within the build context, then they would be deleted at the end of the build. They aren't something that a cluster admin would be expected to manually create or manage. I'll add some clarity around that as well as the why, which is that I'm trying to reduce or eliminate the need to gzip / Base64 encode a rendered MachineConfig into a ConfigMap so that it can be consumed by the build context.

Although perhaps in a post-bootc world, this would be unnecessary in the first place since we could just translate MachineConfigs into ConfigMaps and pass them into bootc.